### ISSUE TITLE
MYSQL to PostgreSQL: "ERROR Argument X is not a NUMBER: #\Nul

### ISSUE THREAD

**Issue by TheSeesaw** [dimitri/pgloader#684]
First of all, thanks for creating pgloader. There's really no other tool like it.

I'm transferring a lot of historical data from an old MYSQL database into a newer PostgreSQL database that holds the same data (no intermediate transformations). I've run the transfer a few times, and have consistently seen the following two errors, which prevent a target row from transferring:

(I've left out table names for anonymity, everything else is accurate from the output)

"ERROR Argument X is not a NUMBER: #\Nul
 ERROR Error while formatting a row from <tablename>"
"ERROR Argument X is not a NUMBER: #\Soh
 ERROR Error while formatting a row from <tablename>"

I've very little searching the internet for this error, so I figured I'd ask here for some guidance. The error message itself is so ambiguous that it's hard pick a direction to proceed in. 

Have you ever seen anything like this?

Is there a logical direction for me to proceed in debugging the process?

Thank you for your time, I look forward to hearing from you soon.

UPDATE: I've just discovered that the error seems to be with the MYSQL bit data type, all fields that are loaded with a non-NULL bit value in MYSQL fail the entire row's transfer to PostgreSQL. Hope this helps.

**Comment by dimitri** [comment]
What is the bit data type casted to in PostgreSQL? I see that the only default casting rule we have for this MySQL data type is for `bit(1)` that is sometimes used as a boolean. What's your use case? I think we need another casting rule and a transformation function for your data.

Other than that, the `#\Nul` is a Common Lisp representation of a Character, which is its own data type in that language. `#\` is followed by the character name. That's pretty useful:

```
CL-USER> (format t "~@c" (aref (the base-string "ï") 0))
#\Latin_Small_Letter_I_With_Diaeresis

CL-USER> (format t "~c ~c" #\Nul #\Soh)
 
```

**Comment by TheSeesaw** [comment]
The bit type ends up being cast to boolean in PostgreSQL. The weird part is that pgloader appears to make the new table definition (with booleans instead of bits) just fine, but doesn't transfer any of the data. The bit(1) is being used in the MYSQL table effectively as a boolean field. I tried using a casting rule that I found on another issue post, but it didn't seem to make a difference. Here's what was in my command.load file:

`load database
   from <source information>
   into <target information>

 CAST type bit when (= precision 1) to boolean drop typemod  using bits-to-boolean

;`

I'm pretty inexperienced with pgloader, so I'm not really sure if this command file is correct. Does that help clarify? 

Thank you for getting back to me, I look forward to hearing from you again soon.

**Comment by dimitri** [comment]
This looks like a bug in pgloader, and it might be due to MySQL having changed the way it returns bit vectors in current versions. All I know is that it used to work when I had MariaDB 5.x here, and I could reproduce your bug with MariaDB 10.2.8 locally. So well, here's a fix that make pgloader accepts either a numeric vector or a character vector as a bit-vector.

Please build from fresh sources and try again, you should be good with the default casting rules.

**Comment by TheSeesaw** [comment]
Okay. So I ran 'make clean' on pgloader, pulled the latest changes and rebuilt it with 'make pgloader'.

I tried running the database transfer again, but I'm still receiving the same error above for the tables with bit fields. I also tried again with the cast above and receive the same errors.

I can verify that the bit(1) is the culprit by setting all of the bit(1) values of one table to NULL, then running pgloader again. This time, the rows that have bit(1) fields set to NULL will transfer successfully where they have failed before.

 Not sure what to do.
